### PR TITLE
K8SPSMDB-332 feature: Restrict running min pods of config servers

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPSMDB-332](https://jira.percona.com/browse/K8SPSMDB-332) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPSMDB-332

Modified `setSafeDefaults` function to restrict running less than 3 pods of configsvr unless allowUnsafeConfigurations flag is set to true by changing configsvr size to 3 if it's lower than that.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |
| demand-backup-sharded | failed |
| data-sharded | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/b4ac127fef1cf109773c74bb22c16895a8a08995
image: `perconalab/percona-server-mongodb-operator:PR-622-b4ac127f`

File: pkg/apis/psmdb/v1/psmdb_defaults.go [hunk 1]
```diff
@@ -476,9 +477,16 @@ func (rs *ReplsetSpec) setSafeDefauts(log logr.Logger) {
 
 	// Replset size can't be 0 or 1.
 	// But 2 + the Arbiter is possible.
-	if rs.Size < 2 {
-		loginfo(fmt.Sprintf("Replset size will be changed from %d to %d due to safe config", rs.Size, defaultMongodSize))
-		rs.Size = defaultMongodSize
+	if rs.ClusterRole == ClusterRoleConfigSvr {
```
Comment (delgod) [inline pkg/apis/psmdb/v1/psmdb_defaults.go:480]: but what if Arbiter is enabled?
in such case, we receive 3 members and 1 arbiter, as result we receive high probability to get split-brain

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | failed |
| upgrade-consistency | passed |
| init-deploy | passed |
| demand-backup | failed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| demand-backup-sharded | failed |
| data-sharded | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/f69f5073535e8874dbb8822b9627bbf1ab2e9186
image: `perconalab/percona-server-mongodb-operator:PR-622-f69f5073`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| arbiter | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| upgrade | passed |
| liveness | passed |
| upgrade-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/91f6c59ce0a422c51c54c8f01d92383ced8405ff
image: `perconalab/percona-server-mongodb-operator:PR-622-91f6c59c`

File: pkg/apis/psmdb/v1/psmdb_defaults.go (hunk 1 above)
Comment (pooknull) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:485]: Should I change `minSafeConfigsvrSize` value to 4 or I should change the arbiter checks logic?

Comment (ihormaltsev) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:485]: @delgod FYI

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| upgrade | failed |
| smart-update | failed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/be30623e9616f52068862a6c079e57eae0e7495b
image: `perconalab/percona-server-mongodb-operator:PR-622-be30623e`

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scheduled-backup | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/056284d9ac0a6c43ee29afa4d948e174dbef3fb6
image: `perconalab/percona-server-mongodb-operator:PR-622-056284d9`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/f9f63633fdfcee750da7f83485b9d45d13de1a7c
image: `perconalab/percona-server-mongodb-operator:PR-622-f9f63633`

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| storage | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/b772460650c564e8bba745e53d9907f1ca4f3c2b
image: `perconalab/percona-server-mongodb-operator:PR-622-b7724606`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| storage | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/622/commits/d6d1bffef3a0a51e59edce9495c54d56e3b1bd62
image: `perconalab/percona-server-mongodb-operator:PR-622-d6d1bffe`

**Pull request merged**